File: kotlin/chatgpt-oauth-core/src/main/kotlin/dev/chatgptoauth/store/MemoryCredentialStore.kt
package dev.chatgptoauth.store

import dev.chatgptoauth.CompareAndSwapResult
import dev.chatgptoauth.CredentialStore
import dev.chatgptoauth.TokenSet
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/** Provides an ephemeral subject-keyed store seeded by [initial] for tests and local development only. */
public class MemoryCredentialStore(initial: Map<String, TokenSet> = emptyMap()) : CredentialStore {
    private val mutex = Mutex()
    private val records = initial.toMutableMap()

    override suspend fun load(subject: String): TokenSet? {
        requireSubject(subject)
        return mutex.withLock { records[subject] }
    }

    override suspend fun compareAndSwap(
        subject: String,
        expectedVersion: Long,
        next: TokenSet,
    ): CompareAndSwapResult {
        requireSubject(subject)
        require(expectedVersion >= 0) { "expectedVersion must be nonnegative." }
        return mutex.withLock {
            val current = records[subject]
            if ((current?.version ?: 0L) != expectedVersion) return@withLock CompareAndSwapResult(false, current)
            val stored = next.copy(version = expectedVersion + 1)
            records[subject] = stored
            CompareAndSwapResult(true, stored)
        }
    }

    override suspend fun delete(subject: String) {
        requireSubject(subject)
        mutex.withLock { records.remove(subject) }
    }

    private fun requireSubject(subject: String) {
        require(subject.isNotBlank()) { "subject must be a nonempty server-derived identifier." }
    }
}

File: swift/Sources/ChatGPTOAuth/Stores/MemoryCredentialStore.swift
import Foundation

/// Stores subject-keyed credentials ephemerally for tests and local development.
public actor MemoryCredentialStore: CredentialStore {
    private var records: [String: TokenSet]

    /// Creates a memory store containing independent value copies of the supplied records.
    public init(initial: [String: TokenSet] = [:]) {
        records = initial
    }

    /// Loads credentials for exactly one explicit subject.
    public func load(subject: String) async throws -> TokenSet? {
        try requireSubject(subject)
        return records[subject]
    }

    /// Replaces one subject only when its current version matches, returning the winning value otherwise.
    public func compareAndSwap(
        subject: String,
        expectedVersion: Int,
        next: TokenSet
    ) async throws -> CompareAndSwapResult {
        try requireSubject(subject)
        let current = records[subject]
        guard (current?.version ?? 0) == expectedVersion else {
            return CompareAndSwapResult(succeeded: false, current: current)
        }

        var stored = next
        stored.version = expectedVersion + 1
        records[subject] = stored
        return CompareAndSwapResult(succeeded: true, current: stored)
    }

    /// Deletes credentials for exactly one explicit subject.
    public func delete(subject: String) async throws {
        try requireSubject(subject)
        records.removeValue(forKey: subject)
    }

    private func requireSubject(_ subject: String) throws {
        guard !subject.isEmpty else {
            throw ChatGPTOAuthError.store(message: "Credential subject must not be empty.")
        }
    }
}
